File: regions/system/library/dart/migration/lib/src/evolution/diff.dart
import 'schema.dart';

sealed class SchemaDiff {}

class TableAddedDiff extends SchemaDiff {
  final Table table;
  TableAddedDiff({required this.table});
}

class TableDroppedDiff extends SchemaDiff {
  final String tableName;
  TableDroppedDiff({required this.tableName});
}

class ColumnAddedDiff extends SchemaDiff {
  final String table;
  final Column column;
  ColumnAddedDiff({required this.table, required this.column});
}

class ColumnDroppedDiff extends SchemaDiff {
  final String table;
  final String columnName;
  ColumnDroppedDiff({required this.table, required this.columnName});
}

class ColumnChangedDiff extends SchemaDiff {
  final String table;
  final String columnName;
  final Column from;
  final Column to;

  ColumnChangedDiff({
    required this.table,
    required this.columnName,
    required this.from,
    required this.to,
  });
}

/// Computes the differences between two schemas.
List<SchemaDiff> diffSchemas(Schema oldSchema, Schema newSchema) {
  final diffs = <SchemaDiff>[];

  final oldTables = {for (final t in oldSchema.tables) t.name: t};
  final newTables = {for (final t in newSchema.tables) t.name: t};

  // Detect dropped tables
  for (final entry in oldTables.entries) {
    if (!newTables.containsKey(entry.key)) {
      diffs.add(TableDroppedDiff(tableName: entry.value.name));
    }
  }

  // Detect added tables and column changes
  for (final entry in newTables.entries) {
    final oldTable = oldTables[entry.key];
    if (oldTable == null) {
      diffs.add(TableAddedDiff(table: entry.value));
    } else {
      _diffColumns(diffs, entry.key, oldTable, entry.value);
    }
  }

  return diffs;
}

void _diffColumns(
    List<SchemaDiff> diffs, String table, Table oldTable, Table newTable) {
  final oldCols = {for (final c in oldTable.columns) c.name: c};
  final newCols = {for (final c in newTable.columns) c.name: c};

  // Detect dropped columns
  for (final colName in oldCols.keys) {
    if (!newCols.containsKey(colName)) {
      diffs.add(ColumnDroppedDiff(table: table, columnName: colName));
    }
  }

  // Detect added and changed columns
  for (final entry in newCols.entries) {
    final oldCol = oldCols[entry.key];
    if (oldCol == null) {
      diffs.add(ColumnAddedDiff(table: table, column: entry.value));
    } else if (oldCol != entry.value) {
      diffs.add(ColumnChangedDiff(
        table: table,
        columnName: entry.key,
        from: oldCol,
        to: entry.value,
      ));
    }
  }
}

File: regions/system/library/go/migration/diff.go
package migration

// SchemaDiffType represents the type of a schema difference.
type SchemaDiffType int

const (
	DiffTableAdded SchemaDiffType = iota
	DiffTableDropped
	DiffColumnAdded
	DiffColumnDropped
	DiffColumnChanged
)

// SchemaDiff represents a single difference between two schemas.
type SchemaDiff struct {
	Type       SchemaDiffType
	Table      string
	Column     *Column // for ColumnAdded
	ColumnName string  // for ColumnDropped/Changed
	From       *Column // for ColumnChanged
	To         *Column // for ColumnChanged
	AddedTable *Table  // for TableAdded
}

// DiffSchemas compares two schemas and returns the differences.
func DiffSchemas(old, new *Schema) []SchemaDiff {
	var diffs []SchemaDiff

	oldTables := make(map[string]*Table)
	for i := range old.Tables {
		oldTables[old.Tables[i].Name] = &old.Tables[i]
	}
	newTables := make(map[string]*Table)
	for i := range new.Tables {
		newTables[new.Tables[i].Name] = &new.Tables[i]
	}

	// Detect dropped tables
	for name := range oldTables {
		if _, ok := newTables[name]; !ok {
			diffs = append(diffs, SchemaDiff{
				Type:  DiffTableDropped,
				Table: name,
			})
		}
	}

	// Detect added tables and column changes
	for name, newTable := range newTables {
		oldTable, ok := oldTables[name]
		if !ok {
			diffs = append(diffs, SchemaDiff{
				Type:       DiffTableAdded,
				Table:      name,
				AddedTable: newTable,
			})
		} else {
			diffs = append(diffs, diffColumns(name, oldTable, newTable)...)
		}
	}

	return diffs
}

func diffColumns(table string, old, new *Table) []SchemaDiff {
	var diffs []SchemaDiff

	oldCols := make(map[string]*Column)
	for i := range old.Columns {
		oldCols[old.Columns[i].Name] = &old.Columns[i]
	}
	newCols := make(map[string]*Column)
	for i := range new.Columns {
		newCols[new.Columns[i].Name] = &new.Columns[i]
	}

	// Detect dropped columns
	for colName := range oldCols {
		if _, ok := newCols[colName]; !ok {
			diffs = append(diffs, SchemaDiff{
				Type:       DiffColumnDropped,
				Table:      table,
				ColumnName: colName,
			})
		}
	}

	// Detect added and changed columns
	for colName, newCol := range newCols {
		oldCol, ok := oldCols[colName]
		if !ok {
			diffs = append(diffs, SchemaDiff{
				Type:   DiffColumnAdded,
				Table:  table,
				Column: newCol,
			})
		} else if !columnsEqual(oldCol, newCol) {
			diffs = append(diffs, SchemaDiff{
				Type:       DiffColumnChanged,
				Table:      table,
				ColumnName: colName,
				From:       oldCol,
				To:         newCol,
			})
		}
	}

	return diffs
}

func columnsEqual(a, b *Column) bool {
	if a.Name != b.Name || a.DataType != b.DataType || a.Nullable != b.Nullable {
		return false
	}
	if a.Default == nil && b.Default == nil {
		return true
	}
	if a.Default == nil || b.Default == nil {
		return false
	}
	return *a.Default == *b.Default
}
